libcloud/compute/drivers/packet.py

```diff
@@ -0,0 +1,254 @@
+
+
+class PacketNodeDriver(NodeDriver):
+    """
+    Packet NodeDriver
+    """
+
+    connectionCls = PacketConnection
+    type = Provider.PACKET
+    name = 'Packet'
+    website = 'http://www.packet.net/'
+
+    NODE_STATE_MAP = {'queued': NodeState.PENDING,
+                      'provisioning': NodeState.PENDING,
+                      'rebuilding': NodeState.PENDING,
+                      'powering_on': NodeState.REBOOTING,
+                      'powering_off': NodeState.REBOOTING,
+                      'rebooting': NodeState.REBOOTING,
+                      'inactive': NodeState.STOPPED,
+                      'deleted': NodeState.TERMINATED,
+                      'deprovisioning': NodeState.TERMINATED,
+                      'failed': NodeState.ERROR,
+                      'active': NodeState.RUNNING}
+
+    def list_nodes(self, location):
+        data = self.connection.request('/projects/%s/devices' % (location),
+                                       params={'include': 'plan'}
+                                       ).object['devices']
+        return list(map(self._to_node, data))
+
+    def list_locations(self):
+        data = self.connection.request('/facilities')\
+            .object['facilities']
+        return list(map(self._to_location, data))
+
+    def list_images(self):
+        data = self.connection.request('/operating-systems')\
+            .object['operating_systems']
+        return list(map(self._to_image, data))
+
+    def list_sizes(self):
+        data = self.connection.request('/plans').object['plans']
+        return list(map(self._to_size, data))
+
+    def create_node(self, project_id, name, size, image, location):
+        """
+        Create a node.
+
+        :return: The newly created node.
+        :rtype: :class:`Node`
+        """
+
+        params = {'hostname': name, 'plan': size.id,
+                  'operating_system': image.id, 'facility': location.id,
+                  'include': 'plan', 'billing_cycle': 'hourly'}
+
+        data = self.connection.request('/projects/%s/devices' % (project_id),
+                                       params=params, method='POST')
+
+        status = data.object.get('status', 'OK')
+        if status == 'ERROR':
+            message = data.object.get('message', None)
+            error_message = data.object.get('error_message', message)
+            raise ValueError('Failed to create node: %s' % (error_message))
+        return self._to_node(data=data.object)
+
+    def reboot_node(self, node):
+        params = {'type': 'reboot'}
+        res = self.connection.request('/devices/%s/actions' % (node.id),
+                                      params=params, method='POST')
+        return res.status == httplib.OK
+
+    def destroy_node(self, node):
+        res = self.connection.request('/devices/%s' % (node.id),
+                                      method='DELETE')
+        return res.status == httplib.OK
+
+    def list_ssh_keys(self):
```

Comment:
Looks like you should be able to implement ssh key management functionality using key pair management methods which are part of the libcloud standard API - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/base.py#L1183

This also means getting rid of `SSHKey` class in favor of `libcloud.compute.base.KeyPair`.

In your case, you can use store id in the `name` attribute and description in the `extra` dict.